src/main/kotlin/org/jmailen/gradle/kotlinter/support/Reporters.kt

```diff
@@ -1,43 +1,60 @@
-import com.pinterest.ktlint.reporter.json.JsonReporter
-import com.pinterest.ktlint.reporter.plain.PlainReporter
-import com.pinterest.ktlint.reporter.sarif.SarifReporter
+import com.pinterest.ktlint.core.ReporterProvider
 import java.io.File
 import java.io.PrintStream
+import java.util.ServiceLoader
 
-/* ktlint-disable enum-entry-name-case */
-internal enum class ReporterType(val fileExtension: String) {
-    checkstyle("xml"),
-    html("html"),
-    json("json"),
-    plain("txt"),
-    sarif("sarif.json"),
+internal enum class ReporterType(val id: String, val fileExtension: String) {
+    Checkstyle(id = "checkstyle", fileExtension = "xml"),
+    Html(id = "html", fileExtension = "html"),
+    Json(id = "json", fileExtension = "json"),
+    Plain(id = "plain", fileExtension = "txt"),
+    Sarif(id = "sarif", fileExtension = "sarif.json"),
+    ;
+
+    companion object {
+
+        fun getById(id: String) =
+            values().firstOrNull { it.id == id } ?: error("Unknown reporter type=$id")
+    }
+}
+
+internal fun reporterPathFor(reporterType: ReporterType, output: File, relativeRoot: File) = when (reporterType) {
+    ReporterType.Checkstyle,
+    ReporterType.Html,
+    ReporterType.Json,
+    ReporterType.Plain,
+    -> output.toRelativeString(base = relativeRoot)
+
+    ReporterType.Sarif -> output.absolutePath
 }
-/* ktlint-enable enum-entry-name-case */
-
-fun reporterFor(reporterName: String, output: File): Reporter {
-    val out = PrintStream(output)
-    return SortedThreadSafeReporterWrapper(
-        when (ReporterType.valueOf(reporterName)) {
-            ReporterType.checkstyle -> CheckStyleReporter(out)
-            ReporterType.html -> HtmlReporter(out)
-            ReporterType.json -> JsonReporter(out)
-            ReporterType.plain -> PlainReporter(out)
-            ReporterType.sarif -> SarifReporter(out)
-        },
-    )
+
+internal fun resolveReporters(
+    enabled: Map<ReporterType, File>,
+): Map<ReporterType, Reporter> {
+    val allReporterProviders = defaultReporters().associateBy { it.id }
+
+    return enabled
+        .filter { (type, _) -> allReporterProviders.containsKey(type.id) }
+        .mapValues { (type, output) ->
+            allReporterProviders.getValue(type.id).get(
+                out = PrintStream(output),
+                opt = type.generateOpt(),
+            )
+        }
 }
 
-fun reporterPathFor(reporter: Reporter, output: File, projectDir: File): String {
-    val unwrappedReporter = (reporter as? SortedThreadSafeReporterWrapper)?.unwrap() ?: reporter
-    return when (unwrappedReporter) {
-        is SarifReporter -> output.absolutePath
-        else -> output.toRelativeString(projectDir)
-    }
+private fun ReporterType.generateOpt() = when (this) {
+    ReporterType.Checkstyle,
+    ReporterType.Html,
+    ReporterType.Json,
+    ReporterType.Sarif,
+    -> emptyMap()
+
+    ReporterType.Plain -> mapOf("color_name" to "DARK_GRAY")
```

Comment:
Ah, here's _interesting_ thing: previously we instantiated reporter directly using its constructor: [source](https://github.com/pinterest/ktlint/blob/2ea18f057bfac03313b4e687302d7148f2eeb5cc/ktlint-reporter-plain/src/main/kotlin/com/pinterest/ktlint/reporter/plain/PlainReporter.kt#L16)
but the ReporterProvider has different defaults: [source](https://github.com/pinterest/ktlint/blob/2ea18f057bfac03313b4e687302d7148f2eeb5cc/ktlint-reporter-plain/src/main/kotlin/com/pinterest/ktlint/reporter/plain/PlainReporterProvider.kt#L16)
I'll report that upstream, but for now I decided to preserve the old behavior. if we wanted we could expose the reporterOptions via extensions and let plugin consumers provide their own options. This might be also related to: https://github.com/jeremymailen/kotlinter-gradle/issues/171 (to pass the `verbose` parameter)